### ISSUE TITLE
Error when trying to execute a chord with an empty group

### ISSUE THREAD

**Issue by george-silva** [celery/celery#3756]
## Steps to reproduce

Hello all!

I have a little process here that constructs a chord, using a group for validation and then a chain of tasks to be executed after the validation, if they all succeed.

The problem is that we get an exception here:

celery.canvas

```python
    def _get_app(self, body=None):
        app = self._app
        if app is None:
            try:
                tasks = self.tasks.tasks  # is a group
            except AttributeError:
                tasks = self.tasks
            app = tasks[0]._app
            if app is None and body is not None:
                app = body._app
        return app if app is not None else current_app

    tasks = getitem_property('kwargs.header', 'Tasks in chord header.')
    body = getitem_property('kwargs.body', 'Body task of chord.')
```

```plain
Traceback (most recent call last):
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/django/core/handlers/exception.py", line 39, in inner
    response = get_response(request)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/rest_framework/viewsets.py", line 83, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/rest_framework/views.py", line 477, in dispatch
    response = self.handle_exception(exc)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/rest_framework/views.py", line 437, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/rest_framework/views.py", line 474, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/george/projetos/tomtom/scout/src/workflow/viewsets.py", line 114, in change
    task = controller.change_to(state)
  File "/home/george/projetos/tomtom/scout/src/workflow/models.py", line 206, in change_to
    return task_runner.run()
  File "/home/george/projetos/tomtom/scout/src/workflow/task_runner.py", line 146, in run
    return job.delay()
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/celery/canvas.py", line 182, in delay
    return self.apply_async(partial_args, partial_kwargs)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/celery/canvas.py", line 1233, in apply_async
    app = self._get_app(body)
  File "/home/george/projetos/.virtualenvs/scout/local/lib/python2.7/site-packages/celery/canvas.py", line 1338, in _get_app
    app = tasks[0]._app
IndexError: list index out of range
```

This happens because there are no tasks in the group.

Here's an excerpt of the code that is failing:

```python
    def run(self):

        '''executes the tasks'''
        cid = self.controller.id
        nsi = self.next.id
        self.controller.inner_state = INNER_STATE_RUNNING
        self.controller.save()
        validation = group([v.s(cid=cid, nsi=nsi)
                            for v in self.validation_tasks])

        tasks = chain([t.s(cid=cid, nsi=nsi)
                       for t in self.tasks])
        if len(validation.tasks) > 0:
            job = chain(validation, tasks)
        else:
            job = tasks
        return job.delay()  # failed here. 
```

I've added the check for the validation group length, and now the problem is gone.

## Expected behavior

Shouldn't Celery skip or properly handle empty groups?

## Actual behavior

Celery fails trying to get the current app.

**Comment by osminogin** [comment]
Got the same issue.

Simple Celery chain silently breaks on empty group (with empty `item_list` in example)

```python
chain(
    group(sending_test.s(campaign, item) for item in item_list),
    sending_aaa.si(campaign),
    sending_bbb.si(campaign)
).delay()
```
My workaround for this situation include empty `skip` task (maybe someone find this helpful):

```python
@shared_task
def skip(): return

chain(
    group(sending_test.s(campaign, item) for item in item_list)
    if len(item_list) else skip.s(),
    sending_aaa.si(campaign),
    sending_bbb.si(campaign)
).delay()
```
But I'm agree that main question is expected Celery behavior.

**Comment by AlexHill** [comment]
@george-silva this was fixed in PR #3847. I'm going to close this, please re-open if you find your issue isn't resolved in the latest master.

@osminogin an empty group as a chord header is a separate issue, but also fixed - see #4443.

**Comment by george-silva** [comment]
@AlexHill I'm pretty sure it's resolved 😄 .

Thanks for acting and the feedback on this.
